mlir/lib/Dialect/MIOpen/utility/transformMapUtils.cpp

```diff
@@ -334,47 +391,52 @@ propagateVectorizationInfo(TransformMapAttr map,
+          result[lower] =
+              VectorizationInfo(lowerMaxLen, input[upper]->needsCoefficient);
         }
       }
       break;
     // The embed rule: as we walk from smaller to larger coefficients, we
-    // accumulate the vectorization coefficient, `maxLength`, initially 1. If a
-    // dimension's coefficient in the embedding is equal to `maxLength`, we
-    // multiply `maxLength` by that dimension's vectorization length and
-    // continue. During this processes, dimensions that have vectorization
-    // information equal to None (those held constant) are ignored entirely.
+    // accumulate the vectorization coefficient by multiplying together the
+    // vectorization lengths of dimensions, stopping if the accumulated length
+    // ends up not equal to theat dimension's coefficient or if the coefficient
+    // isn't equal to the accumulated value times the smallest coefficient seen
+    // on a dimension being vectorized.
     case TransformType::Embed: {
       // Since embed coefficients can go in any order, we need them sorted
       auto &&zip = llvm::zip(params, upperDims);
       SmallVector<std::tuple<int64_t, uint32_t>> data(zip.begin(), zip.end());
       std::sort(data.begin(), data.end());
 
-      Optional<int64_t> maxLength;
+      Optional<VectorizationInfo> ourResult;
       for (auto pair : data) {
         int64_t coefficient = std::get<0>(pair);
         uint32_t upperDim = std::get<1>(pair);
         if (input[upperDim].hasValue()) {
-          if (!maxLength.hasValue())
-            maxLength = 1;
-          int64_t upperLen = input[upperDim].getValue();
-          if (coefficient == maxLength.getValue()) {
-            maxLength = upperLen * (maxLength.getValue());
+          int64_t upperLen = input[upperDim]->maxLength;
+          int64_t needsCoeff = input[upperDim]->needsCoefficient;
+
+          if (!ourResult.hasValue())
+            ourResult = VectorizationInfo(1, coefficient);
+          if (coefficient == needsCoeff &&
+              coefficient ==
+                  (ourResult->maxLength * ourResult->needsCoefficient)) {
+            ourResult->maxLength *= upperLen;
           } else {
             break;
           }
         }
       }
-      result[lowerDims[0]] = maxLength;
+      result[lowerDims[0]] = ourResult;
       break;
     }
     // Like `Embed`, but a bit simpler since we don't have to sort.
-    case TransformType::Unmerge: {
-      Optional<int64_t> maxLength = propagateUnmergeVectorization(
+    case TransformType::Unmerge:
+      result[lowerDims[0]] = propagateUnmergeVectorization(
           llvm::zip(llvm::reverse(upperDims), llvm::reverse(params)), input);
-      result[lowerDims[0]] = maxLength;
       break;
-    }
     // For merges, the input vectorization length is split among the output
     // dimensions, with a dimension getting a vectorization length equal
     // to the gcd of the remaining vectorization length and that dimension's
```

Comment:
Sorry I'm not sure I get it for the merge rule: If the starting vectorization len is 6 and the output dimensions get split to `{3, 4}`, shouldn't the vectorization length be `{1, 4}`?
 - For the 4: since 6 > 4, the last split dimension is split to be fully vectorized
 - For the 1: 6 / 4 = 1, there's not enough vectorization length carried over to the next dimension

Reply:
No, we can't make that split

After every 6 moves in the upper dimension, there may be an arbitrary jump. This means that some of those length-4 "vectors" you propose may not be contiguous.

Consider that our data will end up in
```
x x+1 x+2 x+3 x+4 x+5 y y+1 y+2 y+3 ... y+5
```
when merge makes this 3 x 4 we get
```
x x+1 x+2 x+3
x+4 x+5 y y+1
y+2 y+3 y+4 y+5
```

Note that the second row of that is not a valid vector write.

Reply:
Thank you for the explanation.